rskj-core/src/main/java/co/rsk/peg/lockingcap/LockingCapSupportImpl.java

```diff
@@ -0,0 +1,79 @@
+package co.rsk.peg.lockingcap;
+
+import co.rsk.bitcoinj.core.Coin;
+import co.rsk.peg.lockingcap.constants.LockingCapConstants;
+import co.rsk.peg.vote.AddressBasedAuthorizer;
+import java.util.Optional;
+import org.ethereum.config.blockchain.upgrades.ActivationConfig;
+import org.ethereum.config.blockchain.upgrades.ActivationConfig.ForBlock;
+import org.ethereum.config.blockchain.upgrades.ConsensusRule;
+import org.ethereum.core.SignatureCache;
+import org.ethereum.core.Transaction;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LockingCapSupportImpl implements LockingCapSupport {
+
+    private static final Logger logger = LoggerFactory.getLogger(LockingCapSupportImpl.class);
+    private final LockingCapStorageProvider storageProvider;
+    private final ActivationConfig.ForBlock activations;
+    private final LockingCapConstants constants;
+    private final SignatureCache signatureCache;
+
+    public LockingCapSupportImpl(
+        LockingCapStorageProvider storageProvider,
+        ForBlock activations, LockingCapConstants constants,
+        SignatureCache signatureCache
+    ) {
+        this.storageProvider = storageProvider;
+        this.activations = activations;
+        this.constants = constants;
+        this.signatureCache = signatureCache;
+    }
+
+    @Override
+    public Optional<Coin> getLockingCap() {
+        Optional<Coin> lockingCap = storageProvider.getLockingCap(activations);
+
+        // Before returning the locking cap, check if it was already set
+        if (!lockingCap.isPresent() && activations.isActive(ConsensusRule.RSKIP134)) {
+            // Set the initial Locking Cap value
+            logger.debug("[getLockingCap] {}", "Setting initial Locking Cap value");
+            storageProvider.setLockingCap(constants.getInitialValue());
+        }
+        return storageProvider.getLockingCap(activations);
+    }
+
+    @Override
+    public boolean increaseLockingCap(Transaction tx, Coin newLockingCap) {
+        // Only pre-configured addresses can modify Locking Cap
+        AddressBasedAuthorizer authorizer = constants.getIncreaseAuthorizer();
+        if (!authorizer.isAuthorized(tx, signatureCache)) {
+            logger.warn("[increaseLockingCap] {} {}", "Not authorized address tried to increase Locking Cap. Address: ", tx.getSender(signatureCache));
+            return false;
+        }
+
+        // New Locking Cap must be bigger than current Locking Cap
+        Optional<Coin> currentLockingCap = getLockingCap();
+        if (currentLockingCap.isPresent() && newLockingCap.compareTo(currentLockingCap.get()) < 0) {
+            logger.warn("[increaseLockingCap] {} {}", "Attempted value doesn't increase Locking Cap. Value attempted: ", newLockingCap.value);
+            return false;
+        }
+
+        Coin maxLockingCap = currentLockingCap.get().multiply(constants.getIncrementsMultiplier());
```

Comment:
is there any chance the `get()` will fail?

Reply:
To use `Optional` unveiled different places where we can get `NPE`. In this case, I added a new `IF` statement to ensure we have a current LockingCap value to avoid this NPE. If we don't have a current Locking Cap value that means RSKIP134 is not Active, so we can't set a Locking Cap value.

C.C.: @marcos-iov @apancorb